presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/RowColumnValidator.java

```diff
@@ -0,0 +1,114 @@
+import com.facebook.presto.sql.tree.SingleColumn;
+import com.facebook.presto.verifier.framework.Column;
+import com.google.common.collect.ImmutableList;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.verifier.framework.VerifierUtil.delimitedIdentifier;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.joining;
+
+public class RowColumnValidator
+        implements ColumnValidator
+{
+    @Inject
+    public RowColumnValidator()
+    {
+    }
+
+    @Override
+    public List<SingleColumn> generateChecksumColumns(Column column)
+    {
+        checkColumnType(column);
+
+        Expression checksum = new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(column.getIdentifier()));
+
+        ImmutableList.Builder<SingleColumn> columnsBuilder = ImmutableList.builder();
+        columnsBuilder.add(new SingleColumn(checksum, Optional.of(delimitedIdentifier(getChecksumAlias(column)))));
+
+        for (RowType.Field field : getFields(column)) {
+            Expression fieldChecksum = new FunctionCall(QualifiedName.of("checksum"), ImmutableList.of(
+                    new DereferenceExpression(column.getIdentifier(), new Identifier(field.getName().get()))));
+            columnsBuilder.add(new SingleColumn(fieldChecksum, Optional.of(delimitedIdentifier(getChecksumAlias(column, field)))));
+        }
+
+        return columnsBuilder.build();
+    }
+
+    @Override
+    public ColumnMatchResult validate(Column column, ChecksumResult controlResult, ChecksumResult testResult)
+    {
+        checkColumnType(column);
+
+        ImmutableList<String> aliases = ImmutableList.<String>builder()
+                .add(getChecksumAlias(column))
+                .addAll(getFields(column).stream()
+                        .map(field -> getChecksumAlias(column, field))
+                        .collect(Collectors.toList()))
+                .build();
+
+        return new ColumnMatchResult(
+                aliases.stream().allMatch(alias -> Objects.equals(controlResult.getChecksum(alias), testResult.getChecksum(alias))),
+                prepareMatchMessage(aliases, controlResult, testResult));
+    }
+
+    private static void checkColumnType(Column column)
+    {
+        checkArgument(column.getType() instanceof RowType, "Expect RowType, found %s", column.getType().getDisplayName());
+    }
+
+    private static List<RowType.Field> getFields(Column column)
+    {
+        return ((RowType) column.getType()).getFields();
+    }
+
+    private static String getChecksumAlias(Column column)
+    {
+        return column.getName() + "_checksum";
+    }
+
+    private static String getChecksumAlias(Column column, RowType.Field field)
+    {
+        return column.getName() + "_" + field.getName().get() + "_checksum";
```

Comment:
Use a different delimiter to separate column and fields name for user-friendliness: i.e. `column$field_checksum`

Reply:
Called `Optional.get()` without checking Optional.present().
it is possible that some fields in a `RowType` do not have name, we should somehow reorganize code so that an artificial name is given in case the name of the field is missing.

For example, for a row(a int, varchar, boolean, d double), we can give `varchar` a name of `_col1` and boolean a name `_col2`.

Reply:
It looks like indexing for struct starts from 1. So it should be _col2 for varchar